tawkto/tawkto.php

```diff
@@ -566,6 +582,115 @@ public function migrate_embed_code() {
+			// check current plugin version.
+			$current = self::PLUGIN_VERSION;
+
+			// compare against prev version.
+			$prev = get_option( self::PLUGIN_VERSION_VARIABLE );
+
+			if ( ! empty( $prev ) && version_compare( $prev, $current ) >= 0 ) {
+				// do not do anything.
+				return;
+			}
+
+			// special case: we've never set the version before.
+			// All plugins prior to the current version needs the upgrade.
+			if ( version_compare( $prev, $current ) < 0 ) {
+				// are there upgrade steps depending on how out-of-date?
+				foreach ( self::UPGRADES as $next_version ) {
+					$this->do_upgrade( $next_version );
+					$prev = $next_version;
+				}
+			}
+
+			// update stored plugin version for next time.
+			update_option( self::PLUGIN_VERSION_VARIABLE, $current );
+		}
+
+		/**
+		 * Does the version upgrade depending on the provided plugin version.
+		 *
+		 * @param  string $version Plugin version.
+		 * @return void
+		 */
+		protected function do_upgrade( $version ) {
+			switch ( $version ) {
+				case self::V_0_7_0:
+					$this->upgrade_0_7_0();
+					break;
+			}
+		}
+
+		/**
+		 * Upgrade for version 0.7.0
+		 *
+		 * @return void
+		 */
+		protected function upgrade_0_7_0() {
+			$visibility            = get_option( TawkTo_Settings::TAWK_VISIBILITY_OPTIONS );
+			$included_url_list     = array_map( 'trim', preg_split( '/,/', $visibility['included_url_list'] ) );
+			$excluded_url_list     = array_map( 'trim', preg_split( '/,/', $visibility['excluded_url_list'] ) );
+			$new_included_url_list = array();
+			$new_excluded_url_list = array();
+
+			foreach ( $included_url_list as $included_url ) {
+				$new_included_url_list[] = $included_url;
+
+				$wildcard = PathHelper::get_wildcard();
+
+				if ( CommonHelper::text_ends_with( $included_url, $wildcard ) ) {
+					continue;
+				}
+
+				$new_included_url = rtrim( $included_url, '/' . $wildcard );
+				if ( in_array( $new_included_url, $included_url_list, true ) ) {
+					continue;
+				}
+
+				$new_included_url_list[] = $new_included_url;
+			}
+
+			foreach ( $excluded_url_list as $excluded_url ) {
+				$new_excluded_url_list[] = $excluded_url;
+
+				$wildcard = PathHelper::get_wildcard();
+
+				if ( CommonHelper::text_ends_with( $excluded_url, $wildcard ) === false ) {
+					continue;
+				}
+
+				$new_excluded_url = rtrim( $excluded_url, '/' . $wildcard );
+				if ( in_array( $new_excluded_url, $excluded_url_list, true ) ) {
+					continue;
+				}
+
+				$new_excluded_url_list[] = $new_excluded_url;
+			}
+
+			$visibility['included_url_list'] = join( ', ', $new_included_url_list );
+			$visibility['excluded_url_list'] = join( ', ', $new_excluded_url_list );
+
+			update_option( TawkTo_Settings::TAWK_VISIBILITY_OPTIONS, $visibility );
+		}
+
+		/**
+		 * Registers hooks in admin area
+		 */
+		public function register_hooks_in_admin() {
+			if ( ! is_admin() ) {
+				// do nothing.
+				return;
+			}
+
+			add_action( 'plugins_loaded', array( $this, 'upgrade' ) );
```

Comment:
Tested out the following upgrade hooks (followed how other people did it) but failed to make it work like I wanted to:
- `upgrader_process_completed` - people said that this hook will run once the `update` process of the theme/plugin finishes. Tried it out by setting transient data like `tawkto_updated` flag, like what people did, and checking the flag by using the hook `plugins_loaded` but wasn't able to make it work. The trigger wasn't even called. I checked the DB if the transient data was added and also tried forcing the upgrade script to throw an `Exception` error (which it didn't).
- `upgrader_overwrote_package` - the docs stated that this will get triggered once the plugin package has been overwritten by the updated one. Tried it out a couple of times but the hook also wasn't being triggered.

I did a bit more research but there weren't any answers that were helpful. So I just followed a similar logic to `migrate_embed_code` function but, instead, I used the `plugins_loaded` hook to run the `upgrade` function every time the page is loaded (in this case, every time a page in the admin area is loaded). And the `upgrade` function handles the version checking to decide whether to run through the upgrade or not. (which has a similar logic with the `migrate_embed_code` function).